src/cart/crt_context.c

```diff
@@ -1500,14 +1507,22 @@ crt_context_req_untrack(struct crt_rpc_priv *rpc_priv)
 		crt_rpc_lock(tmp_rpc);
 		D_MUTEX_LOCK(&epi->epi_mutex);
 		if (tmp_rpc->crp_state == RPC_STATE_QUEUED && credits_available(epi) > 0) {
-			tmp_rpc->crp_state = RPC_STATE_INITED;
-			crt_set_timeout(tmp_rpc);
-
-			D_MUTEX_LOCK(&crt_ctx->cc_mutex);
-			rc = crt_req_timeout_track(tmp_rpc);
-			D_MUTEX_UNLOCK(&crt_ctx->cc_mutex);
-			if (rc != 0)
-				RPC_ERROR(tmp_rpc, "crt_req_timeout_track failed, rc: %d.\n", rc);
+			bool submit_rpc = true;
+
+			/* RPC got cancelled or timed out before it got here */
+			if (tmp_rpc->crp_timeout_ts == 0) {
+				submit_rpc = false;
+			} else {
+				tmp_rpc->crp_state = RPC_STATE_INITED;
+				crt_set_timeout(tmp_rpc);
```

Comment:
[Request changes] I'm afraid we've still got a problem. If `submit_rpc` is false, `tmp_rpc` is still in `RPC_STATE_QUEUED`, even as we move it from `waitq` to `req_q`. When `crt_context_req_untrack` works on this RPC later (from `crt_req_timeout_hdlr`), it will decrement `epi_req_wait_num`, which we have already done on new line 1529. I'm wondering if we should instead do

```suggestion
			tmp_rpc->crp_state = RPC_STATE_INITED;
			/* RPC got cancelled or timed out before it got here */
			if (tmp_rpc->crp_timeout_ts == 0) {
				submit_rpc = false;
			} else {
				crt_set_timeout(tmp_rpc);
```

and make the following change to `crt_req_timeout_hdlr`?

```
@@ -1126,7 +1126,9 @@ crt_req_timeout_hdlr(struct crt_rpc_priv *rpc_priv)
 
        switch (rpc_priv->crp_state) {
        case RPC_STATE_QUEUED:
-               RPC_ERROR(rpc_priv, "aborting waiting to group %s, rank %d, tgt_uri %s\n",
+       case RPC_STATE_INITED:
+               RPC_ERROR(rpc_priv, "aborting %s to group %s, rank %d, tgt_uri %s\n",
+                         rpc_priv->crp_state == RPC_STATE_QUEUED ? "waiting" : "inited",
                          grp_priv->gp_pub.cg_grpid, tgt_ep->ep_rank, rpc_priv->crp_tgt_uri);
                crt_context_req_untrack(rpc_priv);
                crt_rpc_complete_and_unlock(rpc_priv, -DER_TIMEDOUT);
```